File: ts/test/markdown.test.ts
/* Copyright (c) 2021-2026 Richard Rodger, MIT License */

import { describe, test } from 'node:test'
import assert from 'node:assert'

import { Tabnas } from '@tabnas/parser'
import { Markdown } from '../dist/markdown'

function md(src: string, opts?: any) {
  return new Tabnas().use(Markdown, opts).parse(src)
}

describe('markdown — prose', () => {
  test('empty and blank', () => {
    assert.deepEqual(md(''), { type: 'document', children: [] })
    assert.deepEqual(md('\n'), { type: 'document', children: [] })
    assert.deepEqual(md('   \n\t\n'), { type: 'document', children: [] })
  })

  test('ATX headings', () => {
    assert.deepEqual(md('# Hello'), {
      type: 'document',
      children: [{ type: 'heading', depth: 1, children: [{ type: 'text', value: 'Hello' }] }],
    })
    assert.deepEqual(md('## Heading 2'), {
      type: 'document',
      children: [{ type: 'heading', depth: 2, children: [{ type: 'text', value: 'Heading 2' }] }],
    })
    assert.deepEqual(md('###### H6'), {
      type: 'document',
      children: [{ type: 'heading', depth: 6, children: [{ type: 'text', value: 'H6' }] }],
    })
    // up to 3 leading spaces, trailing hashes stripped
    assert.deepEqual(md('   ### Hello ###  '), {
      type: 'document',
      children: [{ type: 'heading', depth: 3, children: [{ type: 'text', value: 'Hello' }] }],
    })
    // inline inside heading
    assert.deepEqual(md('# Hello *world*'), {
      type: 'document',
      children: [{
        type: 'heading', depth: 1,
        children: [{ type: 'text', value: 'Hello ' }, { type: 'emphasis', children: [{ type: 'text', value: 'world' }] }]
      }],
    })
  })

  test('Setext headings', () => {
    assert.deepEqual(md('Foo\n==='), {
      type: 'document',
      children: [{ type: 'heading', depth: 1, children: [{ type: 'text', value: 'Foo' }] }],
    })
    assert.deepEqual(md('Bar\n---'), {
      type: 'document',
      children: [{ type: 'heading', depth: 2, children: [{ type: 'text', value: 'Bar' }] }],
    })
    assert.deepEqual(md('Foo\n===\n\nBar\n---'), {
      type: 'document',
      children: [
        { type: 'heading', depth: 1, children: [{ type: 'text', value: 'Foo' }] },
        { type: 'heading', depth: 2, children: [{ type: 'text', value: 'Bar' }] },
      ],
    })
  })

  test('thematic breaks', () => {
    assert.deepEqual(md('---'), { type: 'document', children: [{ type: 'thematicBreak' }] })
    assert.deepEqual(md('***'), { type: 'document', children: [{ type: 'thematicBreak' }] })
    assert.deepEqual(md('___'), { type: 'document', children: [{ type: 'thematicBreak' }] })
    assert.deepEqual(md('  - - -  '), { type: 'document', children: [{ type: 'thematicBreak' }] })
    assert.deepEqual(md('* * *'), { type: 'document', children: [{ type: 'thematicBreak' }] })
  })

  test('paragraphs and soft breaks', () => {
    assert.deepEqual(md('Hello world'), {
      type: 'document',
      children: [{ type: 'paragraph', children: [{ type: 'text', value: 'Hello world' }] }],
    })
    assert.deepEqual(md('Hello\n\nWorld'), {
      type: 'document',
      children: [
        { type: 'paragraph', children: [{ type: 'text', value: 'Hello' }] },
        { type: 'paragraph', children: [{ type: 'text', value: 'World' }] },
      ],
    })
    // paragraph with soft line break inside (joined, inline sees newline as space when breaks:false)
    const doc = md('Paragraph line1\nline2 second')
    assert.equal(doc.children[0].type, 'paragraph')
  })

  test('fenced code', () => {
    assert.deepEqual(md('```\ncode here\n```'), {
      type: 'document',
      children: [{ type: 'code', lang: null, meta: null, value: 'code here' }],
    })
    assert.deepEqual(md('```js\nconsole.log("hi")\n```'), {
      type: 'document',
      children: [{ type: 'code', lang: 'js', meta: null, value: 'console.log("hi")' }],
    })
    assert.deepEqual(md('~~~python\nprint("hi")\n~~~'), {
      type: 'document',
      children: [{ type: 'code', lang: 'python', meta: null, value: 'print("hi")' }],
    })
    assert.deepEqual(md('```js meta info\ncode\n```'), {
      type: 'document',
      children: [{ type: 'code', lang: 'js', meta: 'meta info', value: 'code' }],
    })
  })

  test('indented code', () => {
    assert.deepEqual(md('    indented\n    second'), {
      type: 'document',
      children: [{ type: 'code', lang: null, meta: null, value: 'indented\nsecond' }],
    })
  })

  test('blockquotes', () => {
    assert.deepEqual(md('> hello\n> world'), {
      type: 'document',
      children: [{
        type: 'blockquote',
        children: [{ type: 'paragraph', children: [{ type: 'text', value: 'hello world' }] }]
      }],
    })
    assert.deepEqual(md('> hello\n> - item 1\n> - item 2'), {
      type: 'document',
      children: [{
        type: 'blockquote',
        children: [
          { type: 'paragraph', children: [{ type: 'text', value: 'hello' }] },
          {
            type: 'list', ordered: false, start: null, spread: false,
            children: [
              { type: 'listItem', spread: false, children: [{ type: 'paragraph', children: [{ type: 'text', value: 'item 1' }] }] },
              { type: 'listItem', spread: false, children: [{ type: 'paragraph', children: [{ type: 'text', value: 'item 2' }] }] },
            ]
          }
        ]
      }],
    })
  })

  test('unordered lists', () => {
    const doc = md('- a\n- b\n- c') as any
    assert.equal(doc.children[0].type, 'list')
    assert.equal(doc.children[0].ordered, false)
    assert.equal(doc.children[0].children.length, 3)
    assert.deepEqual(doc.children[0].children[0], {
      type: 'listItem', spread: false,
      children: [{ type: 'paragraph', children: [{ type: 'text', value: 'a' }] }]
    })
  })

  test('ordered lists', () => {
    const doc = md('1. a\n2. b\n3. c') as any
    assert.equal(doc.children[0].ordered, true)
    assert.equal(doc.children[0].start, 1)
    const doc2 = md('5. a\n6. b') as any
    assert.equal(doc2.children[0].start, 5)
  })

  test('html blocks', () => {
    assert.deepEqual(md('<div>\n<p>hi</p>\n</div>'), {
      type: 'document',
      children: [{ type: 'html', value: '<div>\n<p>hi</p>\n</div>' }],
    })
  })

  test('inline — emphasis, strong, code', () => {
    assert.deepEqual(md('Hello *world*'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [
          { type: 'text', value: 'Hello ' },
          { type: 'emphasis', children: [{ type: 'text', value: 'world' }] },
        ]
      }],
    })
    assert.deepEqual(md('Hello **world**'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [
          { type: 'text', value: 'Hello ' },
          { type: 'strong', children: [{ type: 'text', value: 'world' }] },
        ]
      }],
    })
    assert.deepEqual(md('`code`'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{ type: 'inlineCode', value: 'code' }],
      }],
    })
    // foo_bar_baz inside word should NOT be emphasis
    assert.deepEqual(md('foo_bar_baz'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{ type: 'text', value: 'foo_bar_baz' }],
      }],
    })
  })

  test('inline — links and images', () => {
    assert.deepEqual(md('[link](https://example.com)'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{
          type: 'link', url: 'https://example.com', title: null,
          children: [{ type: 'text', value: 'link' }]
        }],
      }],
    })
    assert.deepEqual(md('[link](https://example.com "title")'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{
          type: 'link', url: 'https://example.com', title: 'title',
          children: [{ type: 'text', value: 'link' }]
        }],
      }],
    })
    assert.deepEqual(md('![alt](https://example.com/img.png)'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{ type: 'image', url: 'https://example.com/img.png', title: null, alt: 'alt' }],
      }],
    })
    assert.deepEqual(md('<https://example.com>'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{
          type: 'link', url: 'https://example.com', title: null,
          children: [{ type: 'text', value: 'https://example.com' }]
        }],
      }],
    })
  })

  test('inline — escapes', () => {
    assert.deepEqual(md('Text with \\*escaped\\* stars'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{ type: 'text', value: 'Text with *escaped* stars' }],
      }],
    })
  })

  test('inline — strikethrough gfm', () => {
    assert.deepEqual(md('~~delete~~'), {
      type: 'document',
      children: [{
        type: 'paragraph',
        children: [{ type: 'delete', children: [{ type: 'text', value: 'delete' }] }],
      }],
    })
    const noGfm = new Tabnas().use(Markdown, { gfm: false }).parse('~~delete~~') as any
    assert.equal(noGfm.children[0].children[0].type, 'text')
    assert.equal(noGfm.children[0].children[0].value, '~~delete~~')
  })

  test('mixed document', () => {
    const src = '# Title\n\nParagraph with *em* and **strong**.\n\n- item 1\n- item 2\n\n> quote\n\n```js\ncode\n```'
    const doc = md(src) as any
    assert.equal(doc.children[0].type, 'heading')
    assert.equal(doc.children[1].type, 'paragraph')
    assert.equal(doc.children[2].type, 'list')
    assert.equal(doc.children[3].type, 'blockquote')
    assert.equal(doc.children[4].type, 'code')
  })

  test('bare Tabnas without jsonic', () => {
    // Prose parser must work on the bare engine per dx-report §2.4
    const j = new Tabnas().use(Markdown)
    const doc = j.parse('# hi')
    assert.equal(doc.children[0].type, 'heading')
  })

  test('also works when jsonic already installed (compat)', async () => {
    const { jsonic } = await import('@tabnas/jsonic')
    const j = new Tabnas().use(jsonic).use(Markdown)
    const doc = j.parse('# hi')
    assert.equal(doc.children[0].type, 'heading')
  })
})

File: go/markdown_test.go
package tabnasmarkdown

import (
	"encoding/json"
	"reflect"
	"testing"

	parser "github.com/tabnas/parser/go"
)

func mdParse(src string, opts ...map[string]any) (map[string]any, error) {
	j := parser.Make()
	var o map[string]any
	if len(opts) > 0 {
		o = opts[0]
	}
	if err := j.UseDefaults(Markdown, Defaults, o); err != nil {
		return nil, err
	}
	res, err := j.Parse(src)
	if err != nil {
		return nil, err
	}
	if res == nil {
		return map[string]any{"type": "document", "children": []any{}}, nil
	}
	if m, ok := res.(map[string]any); ok {
		return m, nil
	}
	return nil, nil
}

func jsonRoundMD(v any) any {
	raw, _ := json.Marshal(v)
	var out any
	json.Unmarshal(raw, &out)
	return out
}

func TestEmpty(t *testing.T) {
	res, _ := mdParse("")
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, map[string]any{"type": "document", "children": []any{}}) {
		t.Fatalf("empty: got %#v", got)
	}
	res, _ = mdParse("\n")
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, map[string]any{"type": "document", "children": []any{}}) {
		t.Fatalf("blank: got %#v", got)
	}
}

func TestATXHeading(t *testing.T) {
	res, _ := mdParse("# Hello")
	want := map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "heading", "depth": float64(1), "children": []any{map[string]any{"type": "text", "value": "Hello"}}},
	}}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("ATX: got %#v want %#v", got, want)
	}
	res, _ = mdParse("## Heading 2")
	want = map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "heading", "depth": float64(2), "children": []any{map[string]any{"type": "text", "value": "Heading 2"}}},
	}}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("ATX2: got %#v", got)
	}
}

func TestSetextHeading(t *testing.T) {
	res, _ := mdParse("Foo\n===")
	want := map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "heading", "depth": float64(1), "children": []any{map[string]any{"type": "text", "value": "Foo"}}},
	}}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("setext ===: got %#v want %#v", got, want)
	}
	res, _ = mdParse("Bar\n---")
	want = map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "heading", "depth": float64(2), "children": []any{map[string]any{"type": "text", "value": "Bar"}}},
	}}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("setext ---: got %#v", got)
	}
}

func TestThematicBreak(t *testing.T) {
	res, _ := mdParse("---")
	want := map[string]any{"type": "document", "children": []any{map[string]any{"type": "thematicBreak"}}}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("thematic: got %#v", got)
	}
}

func TestParagraph(t *testing.T) {
	res, _ := mdParse("Hello world")
	want := map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "paragraph", "children": []any{map[string]any{"type": "text", "value": "Hello world"}}},
	}}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("para: got %#v", got)
	}
	res, _ = mdParse("Hello\n\nWorld")
	// two paragraphs
	if m, ok := jsonRoundMD(res).(map[string]any); ok {
		if ch, ok := m["children"].([]any); ok && len(ch) != 2 {
			t.Fatalf("two paras: got %d children", len(ch))
		}
	}
}

func TestFencedCode(t *testing.T) {
	res, _ := mdParse("```\ncode here\n```")
	want := map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "code", "lang": nil, "meta": nil, "value": "code here"}},
	}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("fenced: got %#v want %#v", got, want)
	}
	res, _ = mdParse("```js\nconsole.log(\"hi\")\n```")
	want = map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "code", "lang": "js", "meta": nil, "value": "console.log(\"hi\")"}},
	}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("fenced js: got %#v", got)
	}
}

func TestIndentedCode(t *testing.T) {
	res, _ := mdParse("    indented\n    second")
	want := map[string]any{"type": "document", "children": []any{
		map[string]any{"type": "code", "lang": nil, "meta": nil, "value": "indented\nsecond"}},
	}
	if got := jsonRoundMD(res); !reflect.DeepEqual(got, want) {
		t.Fatalf("indented: got %#v", got)
	}
}

func TestBlockquote(t *testing.T) {
	res, _ := mdParse("> hello\n> world")
	got := jsonRoundMD(res).(map[string]any)
	ch := got["children"].([]any)[0].(map[string]any)
	if ch["type"] != "blockquote" {
		t.Fatalf("blockquote type got %v", ch["type"])
	}
}

func TestLists(t *testing.T) {
	res, _ := mdParse("- a\n- b\n- c")
	got := jsonRoundMD(res).(map[string]any)
	list := got["children"].([]any)[0].(map[string]any)
	if list["ordered"] != false {
		t.Fatalf("unordered: ordered should be false")
	}
	if children, ok := list["children"].([]any); !ok || len(children) != 3 {
		t.Fatalf("unordered: expected 3 items, got %#v", list["children"])
	}
	res, _ = mdParse("1. a\n2. b\n3. c")
	got = jsonRoundMD(res).(map[string]any)
	list = got["children"].([]any)[0].(map[string]any)
	if list["ordered"] != true {
		t.Fatalf("ordered: should be true")
	}
	if list["start"] != float64(1) {
		t.Fatalf("ordered start: got %v", list["start"])
	}
}

func TestInline(t *testing.T) {
	res, _ := mdParse("Hello *world*")
	got := jsonRoundMD(res).(map[string]any)
	para := got["children"].([]any)[0].(map[string]any)
	children := para["children"].([]any)
	if len(children) != 2 || children[1].(map[string]any)["type"] != "emphasis" {
		t.Fatalf("emphasis: got %#v", children)
	}
	res, _ = mdParse("Hello **world**")
	got = jsonRoundMD(res).(map[string]any)
	para = got["children"].([]any)[0].(map[string]any)
	children = para["children"].([]any)
	if children[1].(map[string]any)["type"] != "strong" {
		t.Fatalf("strong: got %#v", children)
	}
	res, _ = mdParse("`code`")
	got = jsonRoundMD(res).(map[string]any)
	para = got["children"].([]any)[0].(map[string]any)
	children = para["children"].([]any)
	if children[0].(map[string]any)["type"] != "inlineCode" {
		t.Fatalf("inlineCode: got %#v", children)
	}
	res, _ = mdParse("foo_bar_baz")
	got = jsonRoundMD(res).(map[string]any)
	para = got["children"].([]any)[0].(map[string]any)
	children = para["children"].([]any)
	if children[0].(map[string]any)["value"] != "foo_bar_baz" {
		t.Fatalf("underscore word: got %#v", children)
	}
}

func TestLinks(t *testing.T) {
	res, _ := mdParse("[link](https://example.com)")
	got := jsonRoundMD(res).(map[string]any)
	para := got["children"].([]any)[0].(map[string]any)
	link := para["children"].([]any)[0].(map[string]any)
	if link["type"] != "link" || link["url"] != "https://example.com" {
		t.Fatalf("link: got %#v", link)
	}
	res, _ = mdParse("![alt](https://example.com/img.png)")
	got = jsonRoundMD(res).(map[string]any)
	para = got["children"].([]any)[0].(map[string]any)
	img := para["children"].([]any)[0].(map[string]any)
	if img["type"] != "image" {
		t.Fatalf("image: got %#v", img)
	}
	res, _ = mdParse("<https://example.com>")
	got = jsonRoundMD(res).(map[string]any)
	para = got["children"].([]any)[0].(map[string]any)
	link = para["children"].([]any)[0].(map[string]any)
	if link["type"] != "link" {
		t.Fatalf("autolink: got %#v", link)
	}
}

func TestStrikethrough(t *testing.T) {
	res, _ := mdParse("~~delete~~")
	got := jsonRoundMD(res).(map[string]any)
	para := got["children"].([]any)[0].(map[string]any)
	del := para["children"].([]any)[0].(map[string]any)
	if del["type"] != "delete" {
		t.Fatalf("delete gfm: got %#v", del)
	}
	res, _ = mdParse("~~delete~~", map[string]any{"gfm": false})
	got = jsonRoundMD(res).(map[string]any)
	para = got["children"].([]any)[0].(map[string]any)
	txt := para["children"].([]any)[0].(map[string]any)
	if txt["type"] != "text" {
		t.Fatalf("delete non-gfm: expected text, got %#v", txt)
	}
}

func TestBareAndCompat(t *testing.T) {
	// Bare engine: the package depends on @tabnas/parser only.
	j := parser.Make()
	if err := j.Use(Markdown, nil); err != nil {
		t.Fatalf("Use Markdown: %v", err)
	}
	res, err := j.Parse("# hi")
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if m, ok := res.(map[string]any); !ok || m["type"] != "document" {
		t.Fatalf("bare: got %#v", res)
	}
}
